internal/pkg/service/stream/storage/level/local/diskwriter/router/connection/connection_test.go

```diff
@@ -0,0 +1,184 @@
+		assert.True(t, conn.IsConnected())
+	}
+	if conn, found := connManager.Connection("w2"); assert.True(t, found) {
+		assert.True(t, conn.IsConnected())
+	}
+
+	// Start another 2 writer nodes
+	w3 := startWriterNode(t, ctx, etcdCfg, "w3")
+	w4 := startWriterNode(t, ctx, etcdCfg, "w4")
+	waitForLog(t, sourceLogger, `{"level":"info","message":"the list of volumes has changed, updating connections","component":"storage.router.connections"}`)
+	waitForLog(t, sourceLogger, `{"level":"info","message":"disk writer client connected from \"%s\" to \"w3\" - \"l%s\"","component":"storage.node.writer.network.client"}`)
+	waitForLog(t, sourceLogger, `{"level":"info","message":"disk writer client connected from \"%s\" to \"w4\" - \"l%s\"","component":"storage.node.writer.network.client"}`)
+	sourceLogger.Truncate()
+	assert.Equal(t, 4, connManager.OpenedConnectionsCount())
+	if conn, found := connManager.Connection("w1"); assert.True(t, found) {
+		assert.True(t, conn.IsConnected())
+	}
+	if conn, found := connManager.Connection("w2"); assert.True(t, found) {
+		assert.True(t, conn.IsConnected())
+	}
+	if conn, found := connManager.Connection("w3"); assert.True(t, found) {
+		assert.True(t, conn.IsConnected())
+	}
+	if conn, found := connManager.Connection("w4"); assert.True(t, found) {
+		assert.True(t, conn.IsConnected())
+	}
+
+	// Shutdown w1 and w3
+	w1.Process().Shutdown(ctx, errors.New("bye bye writer 1"))
+	w1.Process().WaitForShutdown()
+	w3.Process().Shutdown(ctx, errors.New("bye bye writer 3"))
+	w3.Process().WaitForShutdown()
+	waitForLog(t, sourceLogger, `{"level":"info","message":"the list of volumes has changed, updating connections","component":"storage.router.connections"}`)
+	waitForLog(t, sourceLogger, `{"level":"info","message":"disk writer client disconnected from \"w1\" - \"localhost:%s\"","component":"storage.node.writer.network.client"}`)
+	waitForLog(t, sourceLogger, `{"level":"info","message":"disk writer client disconnected from \"w3\" - \"localhost:%s\"","component":"storage.node.writer.network.client"}`)
+	sourceLogger.Truncate()
+	assert.Equal(t, 2, connManager.OpenedConnectionsCount())
+	_, found := connManager.Connection("w1")
+	assert.False(t, found)
+	_, found = connManager.Connection("w3")
+	assert.False(t, found)
+	if conn, found := connManager.Connection("w2"); assert.True(t, found) {
+		assert.True(t, conn.IsConnected())
+	}
+	if conn, found := connManager.Connection("w4"); assert.True(t, found) {
+		assert.True(t, conn.IsConnected())
+	}
+
+	// Shutdown source node - no warning/error is logged
+	s.Process().Shutdown(ctx, errors.New("bye bye source"))
+	s.Process().WaitForShutdown()
+	sourceLogger.AssertJSONMessages(t, `{"level":"info","message":"exited"}`)
+	sourceLogger.Truncate()
+
+	// Shutdown w2 and w4
+	w2.Process().Shutdown(ctx, errors.New("bye bye writer 2"))
+	w2.Process().WaitForShutdown()
+	w4.Process().Shutdown(ctx, errors.New("bye bye writer 4"))
+	w4.Process().WaitForShutdown()
```

Comment:
What if we call `close()` without defer here. Does the shutdown works correctly with calling `close()` of context directly? I would do it for the test, as this Shutdowns are going directly and takes context that was created at the start of the test.

Reply:
cc @jachym-tousek-keboola 

No, it doesn't work that way, it's not designed to be terminated using the context.

Each component has usually its own root context, some wait groups, ...

The component context is cancelled using the `OnShutdown` callback.

`servicectx.Process` is a LIFO of shutdown callbacks.

We don't need to turn off everything at once, but turn it off gradually, in the opposite order of how it started.

Example:
- close etcd session
- THEN close etcd client connection

OR:
- close all stream pipelines
- THEN close all streams in yamux multiplexer (there should be no opened stream)
- THEN close real TCP/KCP connetions (there should be no traffic)

Reply:
When I will have time in future, I would like to do the graceful shutdown using context and waitgroups. However I would close this comment as my expectation was that it is possible to gracefully shutdown the program using `cancel()`.

Reply:
It is graceful shutdown based on ctx and waitgroups, see `OnShutdown` occurences in the repo code.

This is a level higher, the termination of independent components (programs) that have some dependency on each other.

So component == program, in your comment.

------------------

All you have to do is prepare a short example, and I'll be happy to discuss it with you.

Reply:
To add some external reference, the `uber` is doing it in the same way, see `OnStop` hook.
https://github.com/uber-go/fx/blob/8fa92e9e702f362210d2eff5dfea6ac4b9b7a37b/docs/lifecycle.md?plain=1#L42

Reply:
Or https://www.rudderstack.com/blog/implementing-graceful-shutdown-in-go/

**Setup/Shutdown**

There are cases when blocking with ctx code is not the best approach ...